amespahdbpythonsuite/xmlparser.py

```diff
@@ -0,0 +1,525 @@
+                except ValueError:
+                    continue
+                else:
+                    # Isolate metadata.
+                    metadata = children[metadata_index]
+
+                # Additional logic for references tag, if present.
+                if meta_key == 'references':
+                    ref_children = metadata.getchildren()
+                    ref_list = tuple([item.text for item in ref_children])
+                    continue
+
+                # Additional logic for (maybe) multi-element comments tag.
+                if meta_key == 'comments':
+
+                    # Iterate over list of comments, parse individually.
+                    comment_list = []
+                    for _, entry in enumerate(metadata.getchildren()):
+                        assert entry.tag == namespace + 'comment'
+
+                        # key-value pair for comment text.
+                        tmp_dict = {
+                            'comment': entry.text,
+                        }
+
+                        # key-value pair for comment type.
+                        tmp_dict.update(entry.attrib)
+
+                        # store each comment (dict) in the list.
+                        comment_list.append(tmp_dict)
+
+                    continue
+
+                # Extract metadata content.
+                metadata_content = metadata.text
+
+                # Cast to desired type (key dependent).
+                if not isinstance(metadata_content, meta_type):
+                    metadata_content = meta_type(metadata_content)
+
+                # Insert metadata into UID-specific dictionary.
+                sub_dict[meta_key] = metadata_content
+
+            return sub_dict, comment_list, ref_list
+
+        # Unique identifier.
+        try:
+            uid = int(specie.attrib['uid'])
+        except KeyError:
+            uid = 0
+
+        sub_dict, comment_dict, ref_list = \
+            extract_metadata(specie, namespace)
+
+        return uid, sub_dict, comment_dict, ref_list
+
+    @staticmethod
+    def _extract_numeric(specie, namespace, label):
+        """Returns a dictionary (or tuple thereof) for numeric data
+        from an XML 'specie'.
+
+        Args:
+            specie: XML <specie> element.
+            namespace (str): XML namespace.
+            label (str): to identify the element as 'geometry', 'transition',
+                or 'laboratory' data.
+
+        Returns:
+            Tuple of dictionaries. The number of dicts is dependent on
+            what type of data is trying to be extracted: if
+            label == 'laboratory', a single dict containing the
+            frequency/intensity information is present within the
+            tuple. If label is in ['geometry', 'transitions'], one dict
+            per atom in the geometry or vibrational mode is present in
+            the tuple (respectively).
+
+        Note:
```

Comment:
I am a bit unsure on what the best way is to connect specific 'tag-names' with their processing logic. On the one hand, I do like having the tag-names in 'utils/xml_data_types.py'. On the other hand, there seems to be a disconnect, i.e., the XMLParser is not 'standalone/self-contained'.

I've seen examples where the parsing is entirely event-driven, which I suspect is best-practice with the SAP-like-API. That is, instead of explicitly calling functions upon receiving a tag-name-event,  you have similar to what is in 'xml_data_types.py' a dictionary with the tag-name as key and the value a function to be called. Obviously there would be a limit to how deep you would go, e.g., calling a separate function for each 'frequency'-tag would be too much. On the other hand, higher-level tags like 'specie', 'geometry', 'transitions', etc. would be perfectly suited for this approach. Each function would have a only a few if-then-else-statements to recognize the event--the advantage is that events are handled as soon as they come in and not, as now, an enclosing tag has come in. For example, the first level of the iterwalk/parse would simply handle 'comment' and 'species'. On receiving the 'species'-opening tag a function 'parse_species' is called which takes the tree and continues parsing by starting with handling the 'specie'-tags by calling the 'parse_specie' function , which checks handles the 'geometry', 'transitions', etc. tags through appropriate function calls. The parsed data would be stored in a growing dictionary. The species properties could be handled in the 'parse_specie'-function itself. In all cases, when a tag-name is not recognized/in the tag-name/function-dictionary an exception is thrown. In my opinion, the processing/parsing should be more like a 'waterfall'/'tree' as that mimics the structure of the data and the API used. Thinking about adapting the parser for the forthcoming cluster data, this would the be relatively straight-forward, as the only thing that needs to happen is to add an entry in the tag-name/function-dictionary to the appropriate function handling that tag. Note that the only thing the parser in essence does is translating repeated tags into lists of dictionaries and casting the data from string into the proper type.

What are your thoughts?